### PR TITLE
Inference speedups

### PR DESCRIPTION
**Pull request by kislyuk**
This is an initial naive attempt to speed up the inference step by moving as much work as possible into numpy and otherwise vectorizing work where possible. The hotspots to work on were determined by profiling the following command, adapted from test/command_line_tests.sh:

`treetime --aln treetime_examples/data/tb/lee_2015.vcf.gz --vcf-reference treetime_examples/data/tb/tb_ref.fasta --tree treetime_examples/data/tb/lee_2015.nwk --clock-rate 1e-7 --dates treetime_examples/data/tb/lee_2015.metadata.tsv`

Profiling inormation was retrieved like this: `python3 -c'import pstats; p = pstats.Stats("treetime_tottime.cprofile"); p.strip_dirs().sort_stats("tottime").print_stats(42)'`

In my tests, this brings a modest (~10%) but stable speedup of the inference loop.

Relevant to https://github.com/neherlab/treetime/issues/110

cc @sidneymbell @ttung

### CODE REVIEW

**Comment by kislyuk** [comment]
@rneher any advice on getting this reviewed?

**Comment by rneher** [comment]
Thanks, @kislyuk. I'll take a look!

File: treetime/distribution.py [hunk 1]
```diff
@@ -224,14 +225,12 @@ def xmax(self):
...
             valid_idxs = (x > self._xmin-TINY_NUMBER) & (x < self._xmax+TINY_NUMBER)
-            res = np.ones_like (x, dtype=float) * (BIG_NUMBER+self.peak_val)
-            tmp_x = np.copy(x[valid_idxs])
-            tmp_x[tmp_x<self._xmin+TINY_NUMBER] = self._xmin+TINY_NUMBER
-            tmp_x[tmp_x>self._xmax-TINY_NUMBER] = self._xmax-TINY_NUMBER
-            res[valid_idxs] = self._peak_val + self._func(tmp_x)
+            res = np.full(np.shape(x), BIG_NUMBER+self.peak_val, dtype=float)
+            tmp_x = x[valid_idxs]
+            np.clip(tmp_x, self._xmin+TINY_NUMBER, self._xmax-TINY_NUMBER)
+            res[valid_idxs] = self._peak_val + np.interp(tmp_x, self.xvals, self.yvals)
```
Comment (rneher) [inline treetime/distribution.py:233]: what is the advantage of reinstantiating `interp` over the initial instance in line 153 above? it this is faster, one could get rid of the above, right? (while dropping support for anything but linear interpolation)

File: treetime/node_interpolator.py [hunk 2]
```diff
@@ -68,7 +68,7 @@ def _convolution_integrand(t_val, f, g,
             tau = np.unique(np.concatenate((g.x, t_val-f.x,[tau_min,tau_max])))
         else:
             tau = np.unique(np.concatenate((g.x, f.x-t_val,[tau_min,tau_max])))
-        tau = tau[(tau>tau_min-ttconf.TINY_NUMBER)&(tau<tau_max+ttconf.TINY_NUMBER)]
+        np.clip(tau, tau_min-ttconf.TINY_NUMBER, tau_max+ttconf.TINY_NUMBER)
```
Comment (rneher) [inline treetime/node_interpolator.py:71]: did this change lead to a large speed-up? In the previous implementation, `tau` could shrink. Now we might end up with several identical values. what about pulling the `np.unique` out of the `if` statements above and wrapping `np.clip` in the `np.unique`?

**Review by rneher — COMMENTED** [review]
Thanks a lot. This looks mostly good! I made a few small comments/suggestions for additional changes.

**Comment by fanninpm** [comment]
More effective speedups may be found in #121, but that PR has stalled since late July 2020.

**Comment by rneher** [comment]
@kislyuk did you have additional thoughts here? See my review above.

**Comment by rneher** [comment]
> More effective speedups may be found in #121, but that PR has stalled since late July 2020.

this hit a roadblock with numerical instabilities and edge cases. Haven't had time to tackle it comprehensively.

**Comment by rneher** [comment]
I merged this with a few more additions. thanks a lot. 
surprisingly, my local testing suggests that `np.clip` isn't actually faster than `np.maximum(min_val, np.minimum(a, max_val))`, but factored this out in a separate clip function.

**Pull request merged**